### DISCUSSION TITLE
Multicore and multimode port labels for simulation

### DISCUSSION THREAD

**Discussion by simbilod** [gdsfactory/gdsfactory#502] [Ideas]
Hi @joamatab et al,

Currently the tidy3D and MEEP interfaces use hardcoded vertical positions for the source and monitor ports in S-parameter 3D simulations. 

Given the increased availability of multi core layer platforms (e.g. [AMF now offers its silicon nitride layer over silicon core in MPWs](https://opg.optica.org/jlt/abstract.cfm?URI=jlt-39-13-4374)), it would be nice for the simulation plugins to handle ports located in different vertical core layers for 3D simulations (and 2D vertical cross-sections, but this could be later).

This would also be a good time to add options to consider different modes for the same physical port (e.g. TE0, TM0, TE1, etc.) when calculating S-parameters.

Before I implement anything blindly, are there any thoughts on the most sensible way to include this?

Port objects already have a layer property, from which the vertical position could be read according to the layer stack. For the modes, could we add new properties like:

- mode_cutoff (int, default = 1): the maximum mode number to consider, default to only fundamental
- mode_symmetry (None, "TE", or "TM", default=None): which modes to consider, default to lowest regardless of symmetry

**Comment by tvt173** [comment]
@flaport

**Comment by joamatab** [comment]
we could also make sure different port_naming schemes 

for example, we use 'o1', 'o2', 'o3'.... while Floris and Troy use 'in0', 'in1', 'out1' , 'out2'

@HelgeGehring  had a nice idea of using a dict

it would be nice also to standardize the format in which we store the sparameters

What do you think of something like?

| port_in | port_out     | polarization_in     | polarization_out  |
| ------- | ------------ | ------------------- | ----------------- |
| o1      | o2           | te                  | tm                |

**Reply by simbilod** [reply]
Yes, what's important to save on disk and standardize is whatever will be used by the circuit solver downstream to cascade S-parameters. So I would also allow for varying mode order, which is crucial for mode-division multiplexed circuits see e.g. https://arxiv.org/abs/1812.06904 :

| port_in | port_out | pol_in | pol_in | order_in | order_out | 
| --------- | ---------- | -------- | -------- | ---------- | ----------- |
| o1      | o2       | te     | tm     | 0        | 1         | 

This works if the names `o1` and `o2` have well-defined physical locations (including layer).

For within gdsfactory, could we just extend the `class Port(PortPhidl)` with two attributes called `mode_cutoff = int`, and `mode_symmetry = 'all', 'TE', 'TM'`? 

When creating your component, it would look like
```
    c.add_port(
        name="o1",
        midpoint=(-a, 0),
        orientation=180,
        width=width,
        layer=layer_wg,
        cross_section=xs,
        mode_cutoff=1,
        mode_symmetry='TE',
    ) 
```
These can default to `all` and `0` to reproduce the current behaviour on existing components.

Then when the FDTD, EME, etc. get the component port they have all the info they need to calculate S-parameters
